### PR TITLE
4.x: Adjusts CDI OciExtension to use runtime OciExtension for certain authentication tasks

### PR DESCRIPTION
**Pull request by ljnelson**
Per request, this PR adjusts the CDI `OciExtension` to use the runtime `OciExtension` for authentication tasks.

(One thing I'd like feedback on is the change in behavior that will result **in the absence of any configuration**.  In the existing CDI `OciExtension`, in the absence of configuration, a simple `AbstractAuthenticationDetailsProvider` (ADP) implementation will be tried first, followed by a configuration file-based ADP attempt, followed by an "instance principals"-based ADP attempt, followed, finally, by a "resource principal"-based ADP attempt.  (This is [governed by the `enum` order](https://github.com/helidon-io/helidon/blob/707a03bd2b7c334b25fadf9a9bc3cf12da65de01/integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/AdpSelectionStrategy.java#L350-L352) and is currently [documented](https://helidon.io/docs/v3/apidocs/io.helidon.integrations.oci.sdk.cdi/io/helidon/integrations/oci/sdk/cdi/OciExtension.html#configuration:~:text=If%20auto%20is%20present%20in%20the%20list%2C%20or%20if%20no%20value%20for%20this%20property%20exists%2C%20the%20behavior%20will%20be%20as%20if%20config%2Cconfig%2Dfile%2Cinstance%2Dprincipals%2Cresource%2Dprincipal%20were%20supplied%20instead.)).

The runtime `OciExtension` [changes the ordering](https://github.com/helidon-io/helidon/blob/707a03bd2b7c334b25fadf9a9bc3cf12da65de01/integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciAuthenticationDetailsProvider.java#L74-L78).

I'm happy to make any changes that need to be made, or to allow this (maybe?) backwards-incompatible change to occur.  Just let me know.  (If the new ordering is fine, then I'll need to change a lot of Javadoc documentation.)

The PR is simple. The bulk of it simply checks (as before) to see if the user has registered any custom CDI beans for certain pieces (very unlikely, but if she has, we must respect them). (In the "sunny day" case this will of course not happen.)  Then, in this case, which will be the ordinary case, as requested I've changed things so that the runtime `OciExtension`'s `ociAuthenticationProvider` method will be called and its return value will be used.

### CODE REVIEW

**Comment by barchetta** [comment]
The change in ordering looks like a gratuitous incompatibility.  Is there a justification for changing the ordering? I think our principle should be to stay compatible unless we have a justification to break compatibility.

**Comment by ljnelson** [comment]
I see two broad approaches:
1. Change the in-the-absence-of-configuration ordering in the runtime OCI extension to match the existing CDI extension's behavior
2. Somewhat laboriously interrogate the runtime extension's notion of what is available and effectively re-implement its default ordering in the CDI extension to preserve backwards compatibility (so bypass the [in-the-absence-of-configuration `ociAuthenticationProvider()` logic](https://github.com/helidon-io/helidon/blob/e2bd18eadaf398c58eb084113fbf240a1ca51379/integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciExtension.java#L108) (but that may be very method that drove the requesting of this change in the first place?))

I'm happy to make any changes that the community wants.

**Review by tomas-langer — DISMISSED** [review]

**Comment by ljnelson** [comment]
@tomas-langer Thanks for the approval. Do you have any feedback on the behavior change? Or is your feedback: "the behavior change is fine"?

**Comment by ljnelson** [comment]
Pausing this PR per request until changes are made to the "runtime" `OciExtension`.

**Comment by ljnelson** [comment]
Unpausing/undrafting now that, per consensus, I've amended this PR to make requested calls to newly available methods in the runtime OCI extension.

File: integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java [hunk 1]
```diff
@@ -652,14 +665,17 @@ private void processInjectionPoint(@Observes ProcessInjectionPoint<?, ?> event)
...
     private void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager bm) {
-        for (ServiceTaqs serviceTaqs : this.serviceTaqs) {
-            if (serviceTaqs.isEmpty()) {
-                installAdps(event, bm, serviceTaqs.qualifiers());
-            } else {
-                TypeAndQualifiers serviceClientBuilder = serviceTaqs.serviceClientBuilder();
-                TypeAndQualifiers serviceClient = serviceTaqs.serviceClient();
-                installServiceClientBuilder(event, bm, serviceClientBuilder, serviceClient, this.lenientClassloading);
-                installServiceClient(event, bm, serviceClient, serviceTaqs.serviceInterface(), serviceClientBuilder);
+        if (!this.serviceTaqs.isEmpty()) {
```
Comment (trentjeff) [inline integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java:668]: Can we add a test to ensure we are falling back to microprofile config?

Comment (ljnelson) [reply integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java:668]: Like:
```
// in a unit test somewhere
io.helidon.config.Config c = toHelidonConfig(ConfigProvider.getConfig());
OciRuntimeExtension.fallbackConfigSupplier(() -> c);
assertThat(OciRuntimeExtension.configSupplier().get(), is(c));
```
…? Or did you have something else in mind?

Comment (trentjeff) [reply integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java:668]: Yes, put something into the microprofile-config.properties file, and ensure that it got picked up by your code (which will internally defer to OciExtension, etc.)

Comment (ljnelson) [reply integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java:668]: OK…I can do this, but isn't this testing the `io.helidon.integrations.oci.sdk.runtime.OciExtension` class? Are you sure you want the test of how that class should behave in _this_ Java module? There's no place in my code that could regress in this area (I just hand you a converted MicroProfile Config `Config` object and you go from there).

Comment (ljnelson) [reply integrations/oci/sdk/cdi/src/main/java/io/helidon/integrations/oci/sdk/cdi/OciExtension.java:668]: I added a test that hopefully gets at what you're looking for; not sure.

**Pull request merged**